### ISSUE TITLE
Add narrative documentation which demonstrates how to use twisted.names.client module

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6925]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6925|
|Type|enhancement|
|Created|2014-01-21 14:55:47Z|
|Branch|https://github.com/twisted/twisted/tree/names-client-documentation-6925|

There are three decent twisted.names.client examples, but they don't have any accompanying narrative documentation.

There is also some good documentation on [TwistedNames](https://twistedmatrix.com/trac/wiki/TwistedNames), which would be better copied or moved to the main docs directory; where it can be rendered using sphinx for specific versions, and with API links etc.

<details><summary>Searchable metadata</summary>

```
trac-id__6925 6925
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_names_client_documentation_6925 branches/names-client-documentation-6925
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__documentation documentation
time__1390316147000000 1390316147000000
changetime__1390828561000000 1390828561000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

(In [41388]) Branching to 'names-client-documentation-6925'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj commented|
|-|-|

Ready for review in log:branches/names-client-documentation-6925
 * Fixed the twisted.names example tests -- which had broken with the change to sphinx
 * Copied most of jerub's content from [TwistedNames](https://twistedmatrix.com/trac/wiki/TwistedNames) and expanded it a bit
 * Added an example of using the DNS protocols directly -- taken from my Twisted Names talk at PyconUK

I'm not sure whether the use of twisted.conch.stdio is neat or distracting.

It might be worth converting those to self contained example scripts instead.

Build Results:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/names-client-documentation-6925

(Although I don't think the documentation build is actually working with sphinx at the moment.)

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)|@hynek set owner to @hynek|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_4"></a>|@hynek set owner to @wallrj|
|-|-|

Thanks again for working on names documentation; that’s like adding it to Twisted at last.

 1. It builds fine.
 1. Big Picture
  1. Since this is the most basic guide, I figure it could/should be first in the list of guides? (docs/projects/names/howto/index.rst)
  1. I’m not the biggest fan of the repetition within the “In this section:” block. How would you feel about:
      ```
 In this section you will learn:

* about the high level :api:`twisted.names.client <client>` API,
* how you can use the client API interactively from the Python shell (useful for DNS debugging and diagnostics),
* about the :api:`twisted.internet.interfaces.IResolverSimple <IResolverSimple>` and the :api:`twisted.internet.interfaces.IResolver <IResolver>` interfaces,
* about various implementations of those interfaces and when to use them,
* how to customise how the reactor carries out hostname resolution,
* and finally you will also be introduced to some of the low level APIs.
     ```
     That IMHO gives a better idea what will be learned due to less noise.
  1. I wondered what point this `twisted.conch.stdio` thing is about but it apparently runs an reactor so that deferreds fire eventually!? That’s great, TIL. :) Maybe explain it a bit better in line 28? The current logic is that “because conch, we get deferred” but it should be “because conch, deferreds fire eventually”. I’m not sure whether you need to write it out both times though.
  1. Re the Note in 105ff: do you really want it to be three paragraphs? I would suggest to make it a bulleted list since it looks like an enumeration. I would also suggest to link gethostbyname; maybe to http://www.freebsd.org/cgi/man.cgi?query=gethostbyname&section=3 ?
 1. Nits
  1. 21: _T_wisted
  2. 54: POSIX
  3. 69: if you mean generally “resolver” it shouldn’t be capitalized, if you mean some class/interface, it should be `Resolver`’ed. Since you explicitly mention `client.Resolver` in the next line, I would suggest the former.
  4. 93: whenever
  5.  106: POSIX

I would like to see the nits and at least 2.3 and 2.4 addressed, otherwise it looks good to me. Feel free to merge after that whenever you feel confident.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj set status to `assigned`|
|-|-|

Thanks for the review Hynek,

I've addressed all your points and will merge shortly.

More details below:


Replying to [hynek](#note_4):
> Thanks again for working on names documentation; that’s like adding it to Twisted at last.
>  1. It builds fine.
>  1. Big Picture

>   1. Since this is the most basic guide, I figure it could/should be first in
>   the list of guides? (docs/projects/names/howto/index.rst)

r41427

>   1. I’m not the biggest fan of the repetition within the “In this section:” block. How would you feel about:


r41428


>   1. I wondered what point this `twisted.conch.stdio` thing is about but it
>   apparently runs an reactor so that deferreds fire eventually!? That’s great,
>   TIL. :) Maybe explain it a bit better in line 28? The current logic is that
>   “because conch, we get deferred” but it should be “because conch, deferreds
>   fire eventually”. I’m not sure whether you need to write it out both times
>   though.

r41430

>   1. Re the Note in 105ff: do you really want it to be three paragraphs? I
>   would suggest to make it a bulleted list since it looks like an
>   enumeration. I would also suggest to link gethostbyname; maybe to
>   http://www.freebsd.org/cgi/man.cgi?query=gethostbyname&section=3 ?

r41431

I linked to the stdlib documentation instead -- which is what t.i.i.IResolverSimple.getHostByName is really equivalent to.

r41437

>  1. Nits
>   1. 21: _T_wisted

r41432.

>   2. 54: POSIX

r41433

>   3. 69: if you mean generally “resolver” it shouldn’t be capitalized, if you
>   mean some class/interface, it should be `Resolver`’ed. Since you explicitly
>   mention `client.Resolver` in the next line, I would suggest the former.

r41434

>   4. 93: whenever

r41435

>   5.  106: POSIX

r41433


Also thread pool.
r41436

> I would like to see the nits and at least 2.3 and 2.4 addressed, otherwise it
> looks good to me. Feel free to merge after that whenever you feel confident.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj set status to `closed`|
|-|-|

(In [41438]) Merge names-client-documentation-6925

Author: rwall
Reviewers: hynek
Fixes: [#6925](https://github.com/twisted/twisted/issues/6925)

twisted.names now has narrative documentation explaining how to use its client APIs.
